### ISSUE TITLE
Crash on exeption-free release build in luaD_throw (MSVC)

### ISSUE THREAD

**Issue by dbuyakov-gi** [ThePhD/sol2#1011]
Hi,

Unfortunately - another problem with MSVC.
If I try to build example below with argument `/Ob2` (Inline Function Expansion - Any Suitable) and any level of optimization, then programm will crash on calling `longjmp` in `luaD_throw`.

```
#define SOL_NO_EXCEPTIONS 1
#define SOL_SAFE_REFERENCES 1
#define SOL_SAFE_USERTYPE 1
#define SOL_SAFE_FUNCTION_CALLS 1
#define SOL_SAFE_FUNCTION 1
#define SOL_SAFE_NUMERICS 1
#define SOL_SAFE_GETTER 1

#include "sol.hpp"
#include <iostream>

std::string testFunc( std::string str )
{
	return std::string( "hello " ) + str;
}

int main()
{
	sol::state lua;
	lua.open_libraries( sol::lib::base );

	lua["testFunc"] = &testFunc;

	lua.script( R"(
			function f ()
                -- call a function with wrong arguments
                -- to trigger lua_error
                testFunc()
			end
		)" );

	sol::protected_function f( lua["f"] );

	sol::protected_function_result result = f();
	if ( result.valid() )
	{
		// Call succeeded
		std::cout << "call succeeded" << std::endl;
	}
	else
	{
		// Call failed
		sol::error err = result;
		std::string what = err.what();
		std::cout << "call failed, sol::error::what() is " << what << std::endl;
	}
}
```
Worth noting that `/Ob2` and `/O2` are typical flags for release build in Visual Studio.

As far as I tested - this is only reproducible with MSVC (tested on VS2019 and VS2015).
GCC/Clang is free from that "bug" (godbolt examples with high optimation inlining - https://godbolt.org/z/fGd4zf and https://godbolt.org/z/fs3GTs).

**Environment:**
Lua 5.3.5 (build as C-library)
Sol 3.2.1
MSVC compiller (tested with versions 19.00.24215.1 and 19.24.28316)

**Workaround:**
Decrease optimization level to `/Ob1` (Inline Function Expansion - __inline only).

Thank you for this library and thanks in advance for any comments.

**Comment by ThePhD** [comment]
I cannot seem to reproduce this error while building on Visual Studio 2019, x64, in Release mode.

Do you have any other information for me?

**Comment by ThePhD** [comment]
Nevermind, that was debug mode. NOW I go into an unstoppable freeze!

**Comment by ThePhD** [comment]
Okay, I got it to work by doing the following:

- Specifying the flags `/EHc-` and `/EHs-` explicitly.
- Rebuilt everything under RelWithDebInfo and Release builds

That is, I turned off exception handling at the _compiler_ level, since that seems to be what you want. If you were trying to build an exceptionless sol2 with exceptions Enabled in the compiler, uh... I'm not sure what I can do to support that configuration. Lua will always call `longjmp`, and its probable that it will smash your stack in the process of doing so (because `longjmp` does not respect local variables and friends).

Either way, I don't think there's anything I can do to help you here beyond telling you not to use the no-exceptions flag in sol2 unless you mean it.
